### PR TITLE
Add VMC deterministic test for 48 electrons NiO cell.

### PR DESCRIPTION
**Pull request by Hyeondeok-Shin**
## Proposed changes

 This PR is to add VMC deterministic test for 48 electrons NiO cell in order to prepare deterministic tests on large size of supercell.
 References are generated for batched code, legacy CPU and GPU build. Once we confirm this test is stable on the cdash  (~ 5 sec is needed for the test on the workstation. Mixed precision CI could be failed as its tolerance is not adjusted for NiO.), deterministic test for larger NiO cell will follow.

## What type(s) of changes does this code introduce?

- Other (please describe): Add test

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Intel Xeon

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by markdewing** [comment]
test this please

**Comment by prckent** [comment]
I ran this locally and get test failures. The difference are large enough that I wonder if we are using exactly the same reference files?

```
        Start 1878: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1
799/889 Test #1878: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1 ..........................................................   Passed    2.54 sec
        Start 1879: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-kinetic
800/889 Test #1879: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-kinetic ..................................................***Failed    0.05 sec
Tests for series 1
  Testing quantity: Kinetic
    reference mean value     :  207.34064557
    reference error bar      :   0.00000100
    computed  mean value     :  207.41150769
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :   0.07086212  ( 70862.12000002 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail

        Start 1880: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-totenergy
801/889 Test #1880: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-totenergy ................................................***Failed    0.04 sec
Tests for series 1
  Testing quantity: LocalEnergy
    reference mean value     : -371.11474959
    reference error bar      :   0.00000100
    computed  mean value     : -371.11332924
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :   0.00142035  ( 1420.34999999 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail

        Start 1881: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-eeenergy
802/889 Test #1881: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-eeenergy .................................................***Failed    0.04 sec
Tests for series 1
  Testing quantity: ElecElec
    reference mean value     :  76.82246146
    reference error bar      :   0.00000100
    computed  mean value     :  76.76778711
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :  -0.05467435  (-54674.35200001 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail

        Start 1882: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-samples
803/889 Test #1882: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-samples ..................................................   Passed    0.04 sec
        Start 1883: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-potential
804/889 Test #1883: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-potential ................................................***Failed    0.05 sec
Tests for series 1
  Testing quantity: LocalPotential
    reference mean value     : -578.45539516
    reference error bar      :   0.00000100
    computed  mean value     : -578.52483693
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :  -0.06944177  (-69441.77000003 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail

        Start 1884: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-ionion
805/889 Test #1884: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-ionion ...................................................   Passed    0.04 sec
        Start 1885: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-localecp
806/889 Test #1885: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-localecp .................................................***Failed    0.04 sec
06/889 Test #1885: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-localecp .................................................***Failed    0.04 sec
Tests for series 1
  Testing quantity: LocalECP
    reference mean value     : -368.77476518
    reference error bar      :   0.00000100
    computed  mean value     : -368.76879132
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :   0.00597386  ( 5973.85999998 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail

        Start 1886: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-nonlocalecp
807/889 Test #1886: deterministic-NiO_a4_e48_pp-vmc_sdj-1-1-nonlocalecp ..............................................***Failed    0.05 sec
Tests for series 1
  Testing quantity: NonLocalECP
    reference mean value     : -47.20495078
    reference error bar      :   0.00000100
    computed  mean value     : -47.22568980
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :  -0.02073902  (-20739.02400000 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail
```
It is a similar story with the batched tests.

```
sha256sum /scratch/pk7/QMC_DATA_FULL/NiO/NiO-fcc-supertwist111-supershift000-S1.h5 
9f85d6b307540880dda6abada64eaaff65c926baef3f15fb254f6c593d703143  /scratch/pk7/QMC_DATA_FULL/NiO/NiO-fcc-supertwist111-supershift000-S1.h5
```
Configure line
```
cmake -GNinja -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx -DQMC_NIO_MAX_SIZE=64 -DQMC_DATA=/scratch/pk7/QMC_DATA_FULL -DQE_BIN=/scratch/pk7/QMCPACK_CI_BUILDS_DO_NOT_REMOVE/build_gccnew_QE/qe-6.7.0/bin/ -DQMC_COMPLEX=0 ../qmcpack/
```
Today I was using gcc 10.3, OpenBLAS, and openmpi.

```

**Review by prckent — CHANGES_REQUESTED** [review]
(Avoid accidental merge since this will pass in CI)

File: tests/solids/NiO_a4_e48_pp/det_NiO-vmc-short.in.xml [hunk 1]
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0"?>
```
Comment (ye-luo) [inline tests/solids/NiO_a4_e48_pp/det_NiO-vmc-short.in.xml:1]: Remove 'short' from filename

**Comment by prckent** [comment]
If we are using the same files then the most likely explanation is that the Monte Carlo has taken a different path, despite the short length of the run. We can debug this by comparing the initial entries in the scalar files and seeing if the differences build up with time. When we have hit this in the past, by trying a few seed values we have been able to find one that avoids the divergence.

File: tests/solids/NiO_a4_e48_pp/det_NiO-vmc-short.in.xml (hunk 1 above)
Comment (prckent) [reply tests/solids/NiO_a4_e48_pp/det_NiO-vmc-short.in.xml:1]: Agreed

Comment (Hyeondeok-Shin) [reply tests/solids/NiO_a4_e48_pp/det_NiO-vmc-short.in.xml]: Addressed.

**Comment by Hyeondeok-Shin** [comment]
Confirming that h5 file is same and test is passing on full precision real build (CPU) on workstation (intel) and theta (intel), but failing on summit (clang).

Workstation
   Testing quantity: LocalEnergy
     reference mean value     : -371.11474959
     reference error bar      :   0.00000100
     computed  mean value     : -371.11474959
     computed  error bar      :   0.00000000
     pass tolerance           :   0.00000300  (  3.00000000 sigma)
     deviation from reference :   0.00000000  (  0.00000000 sigma)
     error bar of deviation   :   0.00000100
     significance probability :   0.00000000  (gaussian statistics)
     status of this test      :   pass

Theta
  Testing quantity: LocalEnergy
    reference mean value     : -371.11474959
    reference error bar      :   0.00000100
    computed  mean value     : -371.11475153
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :  -0.00000194  ( -1.94000000 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   0.94762045  (gaussian statistics)
    status of this test      :   pass

Summit
  Testing quantity: LocalEnergy
    reference mean value     : -371.11474959
    reference error bar      :   0.00000100
    computed  mean value     : -371.10816169
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :   0.00658790  ( 6587.90000000 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail

Interestingly, it is giving large difference from the first VMC block. It may need to be investigated further with different seeds.

**Comment by prckent** [comment]
Probably Intel compiler+MKL is taking a different route to the other compilations (or vise versa.) Can you please attach the scalar files from full precision real build? I'll also check with Intel locally.

**Comment by Hyeondeok-Shin** [comment]
@prckent Here I attached.

[archive.zip](https://github.com/QMCPACK/qmcpack/files/6744324/archive.zip)

**Comment by ye-luo** [comment]
Potential ways to reduce divergence.
1. Set warmup to 0
2. set time step to much smaller value 1-e3

**Comment by Hyeondeok-Shin** [comment]
As @ye-luo commented, adjusting warmup & time step decrease divergence but its tolerance still seems pretty larger than our default tolerance of 1e-6 for full precision.

warmup 0 & time step 0.001

Workstation
  Testing quantity: LocalEnergy
    reference mean value     : -323.17427571
    reference error bar      :   0.00000100
    computed  mean value     : -323.17427571
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :   0.00000000  (  0.00000000 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   0.00000000  (gaussian statistics)
    status of this test      :   pass

Test status: pass

Summit
  Testing quantity: LocalEnergy
    reference mean value     : -323.17427571
    reference error bar      :   0.00000100
    computed  mean value     : -323.17436996
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :  -0.00009425  (-94.24999996 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail

The smallest divergent I found is 4e-6 at 0.1 time step & 0 warmup

Workstation
Tests for series 1
  Testing quantity: LocalEnergy
    reference mean value     : -370.48506661
    reference error bar      :   0.00000100
    computed  mean value     : -370.48506661
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :   0.00000000  (  0.00000000 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   0.00000000  (gaussian statistics)
    status of this test      :   pass

Test status: pass

Summit

Tests for series 1
  Testing quantity: LocalEnergy
    reference mean value     : -370.48506661
    reference error bar      :   0.00000100
    computed  mean value     : -370.48506267
    computed  error bar      :   0.00000000
    pass tolerance           :   0.00000300  (  3.00000000 sigma)
    deviation from reference :   0.00000394  (  3.94000000 sigma)
    error bar of deviation   :   0.00000100
    significance probability :   0.99991848  (gaussian statistics)
    status of this test      :   fail

Test status: fail

**Comment by prckent** [comment]
Confirming that I have reproduced Intel passing locally.

**Comment by prckent** [comment]
My current plan is to update the reference energies and error bars in this PR based on what I observe from Intel, gcc, llvm and runs using a different seed. (I assume @Hyeondeok-Shin is OK with this).

So far I have seen larger numerical variation than I would *like*, but nothing actually alarming and suggestive of a bug.

For the legacy CUDA we don't have the luxury of knowing what a reasonable variation in the results is or what the correct results are, so if a future HIP build gets a different result that does not mean the problem is with the HIP build.

Interestingly I was able to trigger a crash in the CUDA while experimenting with nrule. Will therefore leave at the default here.

**Comment by prckent** [comment]
I uploaded runs with a new seed value and updated references and error bars. The runs pass locally with gcc, intel, llvm, gcc+cuda. As usual some additional changes might be needed for the CI or cdash but the changes should all be small. The reference data was either intel or gcc+cuda and then error bars were increased to accommodate the observed variability from other compilers, libraries etc. It will be interesting to see what happens with HIP.

While testing a few different seed values I noticed that all the runs had more variability than we are used to from the smaller deterministic tests (both absolute and relative). This is mostly likely due to the larger electron count and energy scale in this system compared to our other test systems.

**Comment by prckent** [comment]
Test this please

**Review by prckent — APPROVED** [review]

**Pull request merged**